### ISSUE TITLE
I tried to reduce the number of latent dimensions, but I still haven't solved this problem. Do you try to solve it?

### ISSUE THREAD

**Issue by star0511** [jameschapman19/cca_zoo#39]
#37 
I tried to reduce the number of latent dimensions, but I still haven't solved this problem. Do you try to solve it?

**Comment by jameschapman19** [comment]
just looking into it now. My windows store has broken with inconvenient timing so couldn't get to your zip archive files but have recreated the problem with some random data of same dimensions

**Comment by star0511** [comment]
I tried to set the values of r and eps to 1e-10 1e-20 1e-30, but there was no solution
you can get the test file here
http://118.195.129.87/myfile/testdata.mat
hope you can help me solve this problem. Thank you

**Comment by jameschapman19** [comment]
Think I've got a fix. Bundling it up with some other changes and then will push and release a new version soon (today)!

**Comment by jameschapman19** [comment]
Latest version (1.7.4) should work for the script/data you shared. I actually think the main problem was that my DeepWrapper makes a validation dataset by default and in this case that validation data had a smaller batch size than number of latent dimensions. DCCA objective almost always ends in nans when (for much the same reason as the linear CCA problem degenerates when n<p). Actually this is something I have been wary of since my (unresolved) issue #13. So the major changes that fix your specific script are that by default .fit() doesn't use validation data. I've also chucked in a clip_value parameter DCCA(clip_value=1) which clips gradients which might help in some case but yours works without.

**Comment by jameschapman19** [comment]
I went ahead and made the above change to the deep API. Small change to your code which I have written here:

`import numpy as np
import scipy.io as sio
from cca_zoo.deepmodels import DCCAE, DeepWrapper
from cca_zoo.deepmodels import objectives, architectures
import torch

filename = "testdata"
data = sio.loadmat(filename)
latent_dims = 2048
device = 'cpu'
encoder_1 = architectures.Encoder(latent_dims=latent_dims, feature_size=2048, layer_sizes=[4096])
encoder_2 = architectures.Encoder(latent_dims=latent_dims, feature_size=2048, layer_sizes=[4096])
decoder_1 = architectures.Decoder(latent_dims=latent_dims, feature_size=2048, layer_sizes=[4096])
decoder_2 = architectures.Decoder(latent_dims=latent_dims, feature_size=2048, layer_sizes=[4096])
# DCCAE
dccae_model = DCCAE(latent_dims=latent_dims, encoders=[encoder_1, encoder_2],
                    decoders=[decoder_1, decoder_2], objective=objectives.CCA, lam=1e-3, r=0, eps=1e-3)
optimizer = torch.optim.Adam(dccae_model.parameters(), lr=1e-3, weight_decay=1e-4)
# hidden_layer_sizes are shown explicitly but these are also the defaults
dccae_model = DeepWrapper(dccae_model, device=device, clip_value=1)
dccae_model.fit((data['XV1'], data['XV2']), epochs=50)`

**Comment by jameschapman19** [comment]
`import numpy as np`

**Comment by jameschapman19** [comment]
spoke too soon got the fail at ~40 epochs :(

**Comment by star0511** [comment]
Do you use the data I provided? I just experimented, and there was something wrong with the second epoch

**Comment by jameschapman19** [comment]
Just by way of update I haven't given up on this just had some things come up. I had used your data but a smaller model I now realise.

**Comment by jameschapman19** [comment]
Have had a bit of luck preventing the nans with some small adjustments in the latest version. Haven't yet run your exact example however. Basically it all comes down to avoiding infinite gradients which is tough when using eigenvalue solvers to numerical precision. My MCCA objective (equivalent to CCA for 2 views) avoids computing eigenvectors explicitly which seems to give some stability benefits as mentioned in the pytorch docs (eig Vs eigvals)

**Comment by jameschapman19** [comment]
This is now running with objective=objectives.MCCA (now the default for DCCAE).

https://colab.research.google.com/drive/1UyW5p961ilMKZFZTa9rlJ1m9ElUeIKye?usp=sharing

**Comment by jameschapman19** [comment]
I've tentatively closed. May come back and do a complete study of how the objectives break down for high dimensional problems.
